### PR TITLE
crypto: Improve memory overhead of export_room_keys by using Vec::retain

### PR DESCRIPTION
**Pull request by andybalaam**
As I understand it, `Vec::retain` should do the same job as `into_iter().filter().collect()` but potentially more efficiently. This could help our problem with running out of memory during export ( https://github.com/matrix-org/matrix-rust-sdk/issues/2914 ).

### CODE REVIEW

File: crates/matrix-sdk-crypto/src/machine.rs [hunk 1]
```diff
@@ -1830,13 +1830,9 @@ impl OlmMachine {
...
-        let sessions: Vec<InboundGroupSession> = self
-            .store()
-            .get_inbound_group_sessions()
-            .await?
-            .into_iter()
-            .filter(|s| predicate(s))
-            .collect();
+        let mut sessions = self.store().get_inbound_group_sessions().await?;
+
+        sessions.retain(|s| predicate(s));
```
Comment (bnjbvr) [inline crates/matrix-sdk-crypto/src/machine.rs:1835]: (uber-nit) Maybe even `sessions.retain(predicate)` would work and be slightly more succinct.

**Review by bnjbvr — APPROVED** [review]
Good catch, thanks :+1: 

It might avoid double-buffering for the entries of `Vec<InboundGroupSession>` indeed; I'd expect one issue is that `get_inbound_group_sessions` will still allocate all the sessions while we're discarding some of them. If `get_inbound_group_sessions` returned an iterator, we could filter out without even allocating the vector, which would be optimal.

File: crates/matrix-sdk-crypto/src/machine.rs (hunk 1 above)
Comment (andybalaam) [reply crates/matrix-sdk-crypto/src/machine.rs]: Nice spot, fixed in a fixup commit.

**Comment by andybalaam** [comment]
> If `get_inbound_group_sessions` returned an iterator

Yes, although I think it would need to be a stream because it's async. I may move on to doing this soonish, if it is needed to fix https://github.com/matrix-org/matrix-rust-sdk/issues/2914

**Pull request merged**